File: Bitwise AND of Numbers Range/Bitwise_AND_of_Numbers_Range.py
# 思路一的py代码

class Solution:
    def rangeBitwiseAnd(self, m: int, n: int) -> int:
        memo = [False] * 31

        for i in range(31):
            if m == n and m & 1 is 1:
                memo[i] = True
            m >>= 1
            n >>= 1

            if m is 0 and n is 0:
                break
        
        res = 0
        for i in range(30, -1, -1):
            res <<= 1
            if memo[i]:
                res += 1
        
        return res
    
# 思路二的py代码

# Runtime: 60 ms, faster than 86.48% of Python3 online submissions for Bitwise AND of Numbers Range.
# Memory Usage: 13.4 MB, less than 21.68% of Python3 online submissions for Bitwise AND of Numbers Range.

class Solution:
    def rangeBitwiseAnd(self, m: int, n: int) -> int:
        
        if m is 0 and n is 0:
            return 0
        
        factor = 0
        while m != n:
            m >>= 1
            n >>= 1
            factor += 1
        
        return m << factor

File: Bitwise AND of Numbers Range/Bitwise_AND_of_Numbers_Range.cpp
// 典型的位运算的题目 解题思路很巧妙
// 思路一，比较低效

// Runtime: 8 ms, faster than 95.68% of C++ online submissions for Bitwise AND of Numbers Range.
// Memory Usage: 8.6 MB, less than 5.45% of C++ online submissions for Bitwise AND of Numbers Range.

class Solution 
{
public:
    int rangeBitwiseAnd(int m, int n) 
    {
        vector<bool> memo(31, false);

        for (int i = 0; i < 31; i++)
        {
            if (m == n && m & 1 == 1)
                memo[i] = true;

            m >>= 1;
            n >>= 1;

            if (m == 0 && n == 0)
                break;
        }

        // 计算最终的结果
        int res = 0;
        for (int i = 30; i >= 0; i--)
        {
            res <<= 1;
            if (memo[i])
                res++;
        }

        return res;
    }
};

// 思路二 相对高效

// Runtime: 8 ms, faster than 95.68% of C++ online submissions for Bitwise AND of Numbers Range.
// Memory Usage: 8 MB, less than 89.81% of C++ online submissions for Bitwise AND of Numbers Range.

class Solution 
{
public:
    int rangeBitwiseAnd(int m, int n) 
    {
        if (m == 0 || n == 0)
            return 0;

        int factor = 0;
        while (m != n)    
        {
            m >>= 1;
            n >>= 1;
            factor++;
        }

        return m << factor;
    }
};